opentelemetry-api/src/opentelemetry/context/__init__.py

```diff
@@ -138,15 +138,96 @@ def __repr__(self):
         asyncio.run(main())
 """
 
+import copy
+import types
+import typing
+
 from .base_context import BaseRuntimeContext
 
-__all__ = ["Context"]
+__all__ = ["BaseRuntimeContext", "Context"]
 
 try:
     from .async_context import AsyncRuntimeContext
 
-    Context = AsyncRuntimeContext()  # type: BaseRuntimeContext
+    _CONTEXT = AsyncRuntimeContext()  # type: BaseRuntimeContext
 except ImportError:
     from .thread_local_context import ThreadLocalRuntimeContext
 
-    Context = ThreadLocalRuntimeContext()
+    _CONTEXT = ThreadLocalRuntimeContext()
+
+
+class Context:
+    def __init__(self) -> None:
+        self.contents = {}  # type: typing.Dict[str, object]
+        self.slot_name = "{}".format(id(self))
+        self._slot = _CONTEXT.register_slot(self.slot_name)
+        self._slot.set(self)
+
+    def __enter__(self) -> "Context":
+        return self
+
+    def __exit__(
+        self,
+        exc_type: typing.Optional[typing.Type[BaseException]],
+        exc_val: typing.Optional[BaseException],
+        exc_tb: typing.Optional[types.TracebackType],
+    ) -> None:
+        pass
+
+    def get(self, key: str) -> "object":
+        return self.contents.get(key)
+
+    @classmethod
+    def value(
+        cls, key: str, context: typing.Optional["Context"] = None
+    ) -> "object":
+        if context is None:
+            return cls.current().contents.get(key)
+        return context.contents.get(key)
+
+    @classmethod
+    def set_value(cls, key: str, value: "object") -> "Context":
+        cls.current().contents[key] = value
+        return cls.snapshot()
+
+    @classmethod
+    def current(cls) -> "Context":
+        if _CONTEXT.current_context is None:
+            ctx = Context()
+            cls.set_current(ctx)
+        return getattr(_CONTEXT, _CONTEXT.current_context.get())  # type: ignore
+
+    @classmethod
+    def snapshot(cls) -> "Context":
+        snapshot = Context()
+        snapshot.contents = cls.current().contents.copy()
+        return snapshot
+
+    @classmethod
+    def set_current(cls, ctx: "Context") -> None:
+        if _CONTEXT.current_context is None:
+            _CONTEXT.current_context = _CONTEXT.register_slot(
+                # change the key here
+                "__current_prop_context__"
+            )
+        _CONTEXT.current_context.set(ctx.slot_name)  # type: ignore
+
+    @classmethod
+    def use(cls, **kwargs: typing.Dict[str, object]) -> typing.Iterator[None]:
+        return _CONTEXT.use(**kwargs)  # type: ignore
+
+    @classmethod
+    def register_slot(
+        cls, name: str, default: "object" = None
+    ) -> "BaseRuntimeContext.Slot":
+        return _CONTEXT.register_slot(name, default)
```

Comment:
I'm slightly confused with this. Here we have a class named `Context` that has a method named `register_slot` that calls `register_slot` on an instance of a different class that exists in this module. So, this seems to be coded like this to make the user always end up registering the slots on the same object, maybe?

But, instead of having a global private module-level `_CONTEXT` object, we can have a `Context` class that in its `__new__` method decides to instantiate an `AsyncRuntimeContext` or a `ThreadLocalRuntimeContext` object depending on the import result. This is more the standard practice for singleton patterns in Python.